[stdlib] Fixed for Cygwin

tinysun212 opened the pull request:
<!-- Please complete this template before creating the pull request. -->
#### What's in this pull request?

<!-- Description about pull request. -->

I submitted PR #2351 which has stdlib patch for Cygwin, but it is not merged since `os(Cygwin)` is not widely accepted. Now the conditional variable CYGWIN can be used for the files of *.swift and *.swift.gyb.

This PR uses CYGWIN instead of the previous unaccepted symbol and has more patch.
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

tinysun212 commented:
@gribozavr, @compnerd, Please review this PR.

gribozavr commented on stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:
```diff
@@ -264,9 +264,26 @@ public enum ProcessTerminationStatus : CustomStringConvertible {
+
+  if waitpid(pid, statusPtr, 0) < 0 {
+    preconditionFailure("waitpid() failed")
+  }
```
Sorry, could you explain how does this work?  Why does the system `waitpid()` function need a Swift box?

tinysun212 replied:
I captured the signature in `/usr/include/sys/wait.h` on Cygwin.
```
/* Allow `int' and `union wait' for the status.  */
typedef union
  {
    int *__int_ptr;
    union wait *__union_wait_ptr;
  } __wait_status_ptr_t  __attribute__ ((__transparent_union__));

pid_t waitpid (pid_t __pid, __wait_status_ptr_t __status, int __options);
```
I'll be happy if someone suggests any simpler way.

gribozavr replied:
```swift
withUnsafePointer(to: status) {
  statusPtr in
  let statusPtrWrapper = __wait_status_ptr_t(__int_ptr: statusPtr)
  waitpid(pid, statusPtrWrapper, 0)
}
```

should work I think.

tinysun212 replied:
It is successfully compiled after 1st line is changed to `withUnsafeMutablePointer(&status) {`.
The compiler taught me how to change. Thank you.

tinysun212 committed:
[stdlib] Fixed for Cygwin

- CYGWIN symbol is used to distinguish Cygwin environment from other OS
  and other environment in Windows.
- Added windows and windowsCygnus to OSVersion in StdlibUnittest

compnerd commented on stdlib/public/Platform/glibc.modulemap.gyb:
```diff
@@ -320,6 +320,13 @@ module SwiftGlibc [system] {
       }
 % end
 
+% if CMAKE_SDK in ["CYGWIN"]:
```
Why limit this to CYGWIN?

tinysun212 replied:
Foundation's `./Foundation/NSFileHandle.swift` uses Glibc's `L_SET`, `L_INCR`, `L_XTND`. In Linux, it is enough only include `unistd.h`, but Cygwin's `unistd.h` doesn't define them. 

To support L_xxx macro, Cygwin should include `sys/file.h` which defines them. But other platforms which already support Foundation would already have the definitions.

compnerd replied:
I think that we can just include module on all targets which have the module.

tinysun212 replied:
OK. I'll change this.

gribozavr commented on stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:
```diff
@@ -264,9 +264,19 @@ public enum ProcessTerminationStatus : CustomStringConvertible {
 public func posixWaitpid(_ pid: pid_t) -> ProcessTerminationStatus {
   var status: CInt = 0
+#if CYGWIN
+  withUnsafeMutablePointer(&status) {
```
Are you sure you don't need the `to:` label?  We recently added it to this function: https://github.com/apple/swift/blob/master/stdlib/public/core/LifetimeManager.swift#L62

tinysun212 replied:
I captured the compiler output. The compiler was patched and built from the snapshot of 7/25.
```
[50/63] Compiling /cygdrive/c/Work/swift/build/Ninja-ReleaseAs...ivate/SwiftPrivateLibcExtras/cygwin/x86_64/SwiftPrivateLibcExtras.
FAILED: cd /cygdrive/c/Work/swift/build/Ninja-ReleaseAssert/swift-cygwin-x86_64/stdlib/private/SwiftPrivateLibcExtras && /usr/bin/cm
. . . . .
/cygdrive/c/Work/swift/swift/stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:268:20: error: extraneous argument label 'to:' in call
  withUnsafePointer(to: status) {
                   ^~~~~
```

gribozavr replied:
This change was made after that snapshot was made.

tinysun212 commented on stdlib/private/SwiftPrivateLibcExtras/SwiftPrivateLibcExtras.swift:
```diff
@@ -97,12 +97,21 @@ public func _stdlib_select(
+#if CYGWIN
+        func asFdSetPtr(
+          _ p: UnsafeMutablePointer<UInt>?
+        ) -> UnsafeMutablePointer<_types_fd_set>? {
```
Updating to 7/29 snapshot, I added new patches.

BTW, I don't know why this is differ to Linux.
The definition part in `sys/select.h` is as follows.
```
typedef struct _types_fd_set {
    fd_mask fds_bits[_howmany(FD_SETSIZE, NFDBITS)];
} _types_fd_set;

#define fd_set _types_fd_set
```
`fd_set` did not work, but `_types_fd_set` worked.

compnerd replied:
Probably because we are dropping the aliasing in the clang import.  Can you please file a SR and send it my way with a tiny test case (something like creating a function which uses `fd_set`)?

tinysun212 replied:
Could you explain some more detail, please?  (dropping aliasing, tiny test case ...)

compnerd replied:
Sure.  The clang-importer will ignore the `#define` based aliasing as its not a real type alias, but rather a token replacement via the preprocessor.  The result is that the module does not contain a definition for `fd_set` although, we should.  We could handle simple token replacements which replace type aliased names with an additional type alias.  This would allow us to use the definition.

After writing out this explanation, Im left wondering, why don't we work around this for now with:
  typealias fd_set = _types_fd_set

gribozavr commented on stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb:
```diff
@@ -1437,6 +1438,8 @@ public enum OSVersion : CustomStringConvertible {
     case .android:
       return "Android"
+    case .windows:
+      return "Windows"
```
For this change to StdlibUnittest, would you mind adding a test like `validation-test/StdlibUnittest/Linux.swift`?

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@tinysun212 Thank you!  The changes LGTM modulo a small request to add a test for StdlibUnittest changes.

tinysun212 commented:
@gribozavr I changed for StdlibUnittest.

CodaFi commented:
Dmitri is on vacation for a while.  In the meantime, @swift-ci please test.

Is there a reason you called this `windowsCygnus` and not `cygwin` or `windowsCygwin`?

CodaFi commented:
@swift-ci please test OS X platform.

tinysun212 commented:
It is only for the clear expression that Cygwin is related to Windows.
`windowsCygnus` is from the llvm target string, `*-*-windows-cygnus`.

I'm preparing a PR for MinGW, it will have `windowsGnu` which is analogous to `*-*-windows-gnu`, and they evaluate the TestRunPredicate `.windowsAny` to be true .

CodaFi commented:
Ah, thanks. LGTM.  

@swift-ci please test.

CodaFi commented:
@tinysun212 It looks like that #2351 is still workable and would make this patch a lot cleaner.  Let's see about getting that merged.

tinysun212 commented:
`validation-test/StdlibUnittest/Cygwin.swift` was failed and patched.

CodaFi commented:
@swift-ci please test.

CodaFi commented:
@swift-ci please smoke test.

CodaFi commented:
Ok...

@swift-ci please smoke test OS X platform.

tinysun212 commented:
@compnerd I applied your review, thanks.
@CodaFi Now I applied all of the reviews.

compnerd commented on validation-test/StdlibUnittest/ChildProcessShutdown/FailIfChildCrashesDuringShutdown.swift:
```diff
@@ -2,7 +2,7 @@
 
 import StdlibUnittest
-#if os(Linux) || os(FreeBSD) || os(PS4) || os(Android)
+#if os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Windows)
```
Doesn't this break on MSVC hosts?

tinysun212 replied:
Not tested on windows-msvc. Only tested on Cygwin, and I believe it will be pass for MinGW.

Did you build an alternatives to Glibc for MSVC ?

compnerd replied:
Yes, msvcrt :-).  Its a swift wrapper for ucrt and visualc.

tinysun212 replied:
Would you change after this PR?
I think the code would be like this.
```
#if os(Windows) && MSVC
import Ucrt
import VisualC
#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Windows)
import Glibc
#else
import Darwin
#endif

tinysun212 replied:
@compnerd Could you share the build script (or manual) for MSVC ?

compnerd replied:
I build on Linux.  Just set INCLUDE and LIB as per Windows (adjust them for the movement from C:\ to whatever you're using) and add WINDOWS to the targets to build.  I can give you my cmake/ninja invocation later when I'm near a computer.

compnerd replied:
I just do the following:

add `-DSWIFT_SDKS="LINUX;WINDOWS" -DCMAKE_AR=$(which llvm-ar) -DCMAKE_RANLIB=$(which llvm-ranlib) -DSWIFT_WINDOWS_ICU_I18N_INCLUDE=${SWIFT_TOP_SRCDIR}/build/icu/source/i18n -DSWIFT_WINDOWS_ICU_UC_INCLUDE=${SWIFT_TOP_SRCDIR}/build/icu/source/common` via `--extra-cmake-options`

You need to make sure that ld.lld is available to your toolchain, as building for Windows (MSVC) on !Windows will enable lld.

Additionally, I have the following environment variables setup:
INCLUDE='${SDK_TOP_SRCDIR}/Microsoft Visual Studio 14.0/VC/include;${SDK_TOP_SRCDIR}/Windows Kits/10/Include/10.0.10586.0/ucrt;${SDK_TOP_SRCDIR}/Windows Kits/10/Include/10.0.10586.0/um;${SDK_TOP_SRCDIR}/Windows Kits/10/Include/10.0.10586.0/shared'
LIB='${SDK_TOP_SRCDIR}/Microsoft Visual Studio 14.0/VC/lib;${SDK_TOP_SRCDIR}/Windows Kits/10/Lib/10.0.10586.0/ucrt/x86;${SDK_TOP_SRCDIR}/Windows Kits/10/Lib/10.0.10586.0/um/x86'

Within those paths which are directly lifted from a windows SDK installation, I have two symlinks constructed: ${SDK_TOP_SRCDIR}/Microsoft Visual Studio 14.0/VC/include/module.map -> ${SWIFT_TOP_SRCDIR}/swift/stdlib/public/Platform/visualc.modulemap
${SDK_TOP_SRCDIR}/Windows Kits/10/Include/10.0.10586.0/ucrt -> ${SWIFT_TOP_SRCDIR}/swift/stdlib/public/Platform/ucrt.modulemap

The import libraries for ICU are in the build directory as well.  That tends to be sufficient to cross-compile from Linux x86_64 to Windows x86.

tinysun212 replied:
Thanks you.
Did you use the swift/clang compiler for Linux ?
I'll run those on my Linux machine.

compnerd replied:
Yeah, Ive been doing the entire build on Linux (and should be pretty trivial to reproduce on Darwin).

compnerd replied:
This still breaks the non-cygwin case :/  You never addressed this!

CodaFi commented:
In the meantime @tinysun212, could you please rebase again?

CodaFi commented:
@swift-ci please smoke test macOS platform.

tinysun212 commented:
Rebased and incorporates some of @kstaring PR #4804

moiseev commented:
@swift-ci Please test

tinysun212 commented:
In the building console of Swift Test OS X Platform, there is a liking error for `Benchmark_Onone-arm64-apple-ios8.0`.
Does it caused by this changes?

moiseev commented:
@tinysun212 I don't think this change is responsible for the failure. Let's try again.
@swift-ci Please smoke test macOS platform

moiseev commented:
@swift-ci Please smoke test macOS platform

moiseev commented:
@swift-ci Please smoke test macOS platform

tinysun212 commented:
I found a strange message in the CI test log.

`pull request #35 from tinysun212/pr-cygwin-1`, what is this?, I don't have the PR of number 35.

```
11:21:11 + git fetch --recurse-submodules=yes
11:21:12 + git clean -fdx
11:21:12 + git submodule foreach --recursive git clean -fdx
11:21:12 + git submodule foreach --recursive git reset --hard HEAD
11:21:12 + git reset --hard HEAD
11:21:12 HEAD is now at c324ee3 Merge pull request #35 from tinysun212/pr-cygwin-1
11:21:12 + git checkout master
11:21:12 Already on 'master'
11:21:12 Your branch is up-to-date with 'origin/master'.
11:21:12 + git reset --hard origin/master
11:21:12 HEAD is now at c324ee3 Merge pull request #35 from tinysun212/pr-cygwin-1
11:21:12 + git fetch --recurse-submodules=yes
11:21:12 + git clean -fdx
11:21:13 + git submodule foreach --recursive git clean -fdx
11:21:13 + git submodule foreach --recursive git reset --hard HEAD
11:21:13 + git reset --hard HEAD
11:21:13 HEAD is now at 5af77f3 Merge pull request #95 from kainjow/master
```

tinysun212 commented:
Anyway, I rebased to most recent commit.
I tested on macOS with `build-script.sh -R -t`, couldn't see any failed message.
Please test again ?

shahmishal commented:
@tinysun212 Do you have the CI build link where you saw this message?

tinysun212 commented:
Yes, [here](https://ci.swift.org/job/swift-PR-osx/3949/consoleFull#-1586822805ee1a197b-acac-4b17-83cf-a53b95139a76).

shahmishal commented:
@tinysun212 this came from ToT llbuild repo, looks like this was merged recently. 

https://github.com/apple/swift-llbuild/pull/35

tinysun212 commented:
That was my PR, I can remember now, I'm sorry I forgot it. The CI build log displayed it under the line
`From https://github.com/apple/swift-llvm`, so I couldn't think about that.

Although that was not the cause of the fail of tests, please smoke test again?

shahmishal commented:
@swift-ci Please smoke test

tinysun212 commented:
This time, the error is caused by duplicating glibc modulemap changes with #5131 which was merged 1 hour ago. I'll soon resolve this.

tinysun212 commented:
Resolving the duplicate, I rebased and merged to one commit.

tinysun212 commented:
please test

gparker42 commented:
@swift-ci Please test

slavapestov commented:
@tinysun212 Sorry for dropping the ball on this PR. Do you mind rebasing and fixing the merge conflicts so that we can get this merged in?

tinysun212 commented:
@slavapestov I will resolve this PR this weekend.

slavapestov commented:
Thanks!

slavapestov commented:
@swift-ci Please test

compnerd commented on validation-test/StdlibUnittest/ChildProcessShutdown/FailIfChildExitsDuringShutdown.swift:
```diff
@@ -2,7 +2,7 @@
 
 import StdlibUnittest
-#if os(Linux) || os(FreeBSD) || os(PS4) || os(Android)
+#if os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Windows)
```
Doesnt `os(Windows)` evaluate true for MSVC builds?  This breaks Windows.

tinysun212 replied:
os(Windows) is true for MSVC, MinGW, Cygwin.

compnerd commented on validation-test/StdlibUnittest/ChildProcessShutdown/PassIfChildCrashedDuringTestExecution.swift:
```diff
@@ -2,7 +2,7 @@
 
 import StdlibUnittest
-#if os(Linux) || os(FreeBSD) || os(PS4) || os(Android)
+#if os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Windows)
```
Same.

compnerd commented on validation-test/StdlibUnittest/Stdin.swift:
```diff
@@ -6,7 +6,7 @@ import StdlibUnittest
 #if os(OSX) || os(iOS) || os(tvOS) || os(watchOS)
 import Darwin
-#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android)
+#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Windows)
```
And here
